### PR TITLE
net/http: Preserve some features of headers on the wire

### PR DESCRIPTION
**Pull request by cjpatton**
Adds a field to net/http.Request that allows us to reconstruct the HTTP headers (almost) as they appeared on the wire.

### CODE REVIEW

File: src/net/textproto/reader.go [hunk 1]
```diff
@@ -490,28 +495,31 @@ func (r *Reader) ReadMIMEHeader() (MIMEHeader, error) {
 		strs = make([]string, hint)
 	}
 
+	ordered := make([]CFHeaderLine, 0, hint)
```
Comment (Lekensteyn) [inline src/net/textproto/reader.go]: Can you make all of this conditional via a flag in `http.Server`? If this information is not needed, then these allocations can be skipped. This also avoids the need to set values to `nil` during `request_test.go` as well as all those other tests.

**Review by Lekensteyn — COMMENTED** [review]
Looks fine for the most part except for enabling the collection by default which affects other tests.

File: src/net/textproto/reader.go (hunk 1 above)
Comment (cjpatton) [reply src/net/textproto/reader.go]: Done.

File: src/net/http/request.go [hunk 2]
```diff
@@ -1071,11 +1080,14 @@ func readRequest(b *bufio.Reader, deleteHostHeader bool) (req *Request, err erro
 	}
 
 	// Subsequent lines: Key: value.
-	mimeHeader, err := tp.ReadMIMEHeader()
+	mimeHeader, orderedHeader, err := tp.CFReadMIMEHeader()
 	if err != nil {
 		return nil, err
 	}
 	req.Header = Header(mimeHeader)
+	if cfRecordRequestLines {
```
Comment (Lekensteyn) [inline src/net/http/request.go]: The global `cfRecordRequestLines` variable has to be removed and it needs to somehow use the `net/http.Server` field.

Also just set `CFHeaderLines` unconditionally here `CFReadMIMEHeader` *should* return `nil` when the feature is not enabled.

File: src/net/textproto/reader.go [hunk 3]
```diff
@@ -481,6 +481,15 @@ func (r *Reader) ReadDotLines() ([]string, error) {
 //	}
 //
 func (r *Reader) ReadMIMEHeader() (MIMEHeader, error) {
+	header, _, err := r.readMIMEHeader(false /* cfRecordRequestLines */)
+	return header, err
+}
+
+func (r *Reader) CFReadMIMEHeader() (MIMEHeader, []CFHeaderLine, error) {
+	return r.readMIMEHeader(true /* cfRecordRequestLInes */)
```
Comment (Lekensteyn) [inline src/net/textproto/reader.go]: `I` in `LInes is accidentally capitalized. However I believe that you have to expose this bool parameter anyway (and set it to the http.Server parameter).

**Review by Lekensteyn — CHANGES_REQUESTED** [review]

File: src/net/http/request.go (hunk 2 above)
Comment (cjpatton) [reply src/net/http/request.go]: `cfRecordRequestLines` isn't a global variable here, it's a parameter of function `readRequest`. The parameter *is* set to a global variable when called by `ReadRequest`, but it's set by `Server.CFRecordRequestLines` when called by the server.

File: src/net/textproto/reader.go (hunk 3 above)
Comment (cjpatton) [reply src/net/textproto/reader.go]: Added `recordHeaderLInes bool` parameter to `CFReadMIMEHeader`. In net/http this is set to `Server.CFRecordRecordLInes`.

**Review by wbl — DISMISSED** [review]

File: src/net/http/request.go (hunk 2 above)
Comment (Lekensteyn) [reply src/net/http/request.go]: > The parameter is set to a global variable when called by ReadRequest

Yes that is exactly what I mean, you introduce a new global:
```diff
 // only supports HTTP/1.x requests. For HTTP/2, use golang.org/x/net/http2.
 func ReadRequest(b *bufio.Reader) (*Request, error) {
-       return readRequest(b, deleteHostHeader)
+       return readRequest(b, deleteHostHeader, cfRecordRequestLines)
 }

-// Constants for readRequest's deleteHostHeader parameter.
+// Constants for readRequest's flags.
 const (
        deleteHostHeader     = true
        keepHostHeader       = false
+       cfRecordRequestLines = false
 )
```

However after looking at it again, its sole purpose appears to be documentation. In `server.go`, `conn.readRequest` does propagate the `Server.CFRecordRequestLines` flag:

```diff
-       req, err := readRequest(c.bufr, keepHostHeader)
+       req, err := readRequest(c.bufr, keepHostHeader, c.server.CFRecordRequestLines)
        if err != nil {
```

Consider this concern resolved.

File: src/net/textproto/reader.go [hunk 4]
```diff
@@ -481,6 +481,15 @@ func (r *Reader) ReadDotLines() ([]string, error) {
 //	}
 //
 func (r *Reader) ReadMIMEHeader() (MIMEHeader, error) {
+	header, _, err := r.readMIMEHeader(false /* recordRequestLines */)
+	return header, err
+}
+
+func (r *Reader) CFReadMIMEHeader(recordRequestLines bool) (MIMEHeader, []CFHeaderLine, error) {
+	return r.readMIMEHeader(recordRequestLines)
```
Comment (Lekensteyn) [inline src/net/textproto/reader.go]: This function can be removed now and the other function below can be renamed to `CFReadMIMEHeader` right?

**Review by Lekensteyn — DISMISSED** [review]
LGTM, one minor nit

File: src/net/textproto/reader.go (hunk 4 above)
Comment (cjpatton) [reply src/net/textproto/reader.go]: Done.

**Review by Lekensteyn — APPROVED** [review]

**Review by wbl — APPROVED** [review]

File: src/net/http/request.go [hunk 5]
```diff
@@ -1009,16 +1017,27 @@ func putTextprotoReader(r *textproto.Reader) {
 // requests and handle them via the Handler interface. ReadRequest
 // only supports HTTP/1.x requests. For HTTP/2, use golang.org/x/net/http2.
 func ReadRequest(b *bufio.Reader) (*Request, error) {
-	return readRequest(b, deleteHostHeader)
+	return readRequest(b, deleteHostHeader, cfRecordRequestLines)
+}
+
+// CFReadREquest is like ReadRequest except that it sets CFHeaderLines in the
```
Comment (Lekensteyn) [inline src/net/http/request.go]:
```suggestion
// CFReadRequest is like ReadRequest except that it sets CFHeaderLines in the
```

**Review by Lekensteyn — APPROVED** [review]
small spelling nit. Other than that exposing this flag for simplification of tests sounds reasonable. The alternative (tests requiring sending manual TCP payloads and a special HTTP Handler) is also possible but slightly more complicated.

**Pull request merged**